### ISSUE TITLE
False positive for `inconsistent-return-statements` with method annotated with `NoReturn` return type

### ISSUE THREAD

**Issue by kdestin** [pylint-dev/pylint#8747]
### Bug description

`repro.py`
```python
from typing import NoReturn

class Foo:
    def _never_returns() -> NoReturn:
        raise Exception

    def bar(self) -> int:
        try:
            return 1
        except Exception:
            self._never_returns()
```

### Configuration

```ini
Default Config
```

### Command used

```shell
pylint --disable=all --enable=inconsistent-return-statements repro.py
```

### Pylint output

```shell
************* Module repro
repro.py:7:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)

------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 0.00/10, +8.89)
```

### Expected behavior

```console
------------------------------------
Your code has been rated at 10.00/10
```

### Pylint version

```shell
My pylint configuration is pinned to 


pylint 2.15.8
astroid 2.13.5
Python 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]


But I was able to repro this on the latest released version of pylint (2.17.4 at time of writing).
```

### OS / Environment

```console
$ lsb_release -a
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.2 LTS
Release:	22.04
Codename:	jammy
$ uname -r
5.15.0-72-generic
```

---

I did an initial investigation, and _might_ have an idea of what's going on:

* During the `inconsistent-return-statements` lint, there's an attempt made infer whether the return type of a function call is `NoReturn`.

  https://github.com/pylint-dev/pylint/blob/a67ffbdd261dffa01045162506614bd8c1e21be5/pylint/checkers/refactoring/refactoring_checker.py#L2001-L2016

* But `_is_function_def_never_returning` only inspects the return type if the node passed in is a `nodes.FunctionDef`. But calling `node.func.inferred()[0]` on the method call (e.g. `self._never_returns()`) returns the corresponding method definition node as type `astroid.BoundMethod`. Since that isn't a `nodes.FunctionDef`, `_is_function_def_never_returning` never even looks to see if the return type is `NoReturn`.

  https://github.com/pylint-dev/pylint/blob/a67ffbdd261dffa01045162506614bd8c1e21be5/pylint/checkers/refactoring/refactoring_checker.py#L1956-L1957
